File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
 
    int getHeight(TreeNode* node_ptr) {
        if (node_ptr == NULL) return 0;
        
        int left_node = getHeight(node_ptr->left),
            right_node = getHeight(node_ptr->right);
        return max(left_node, right_node) + 1;
    }

    bool isBalanced(TreeNode* root) {
        if (root == NULL) return true;

        if (abs(getHeight(root->left) - getHeight(root->right)) > 1)
            return false;
        return isBalanced(root->left) && isBalanced(root->right);
    }
};

File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
        	return 0
        return 1 + self.helper(root)


    def helper(self, root):
    	if not root.left and not root.right:
    		return 0
    	if root.left and root.right:
    		return max(self.helper(root.left), self.helper(root.right)) + 1
    	if root.left:
    		return self.helper(root.left) + 1
    	if root.right:
    		return self.helper(root.right) + 1